### ISSUE TITLE
ERL-193: tcp_inet port is not closed when socket is closed in OS, effectively leaking memory

### ISSUE THREAD

**Issue by OTP-Maintainer** [erlang/otp#3383]
Original reporter: `jakub`
Affected versions: `OTP-18.3`, `R16B03-1`
Fixed in versions: `OTP-20.0.1`, `OTP-19.3.6.1`
Component: `erts`
Migrated from: https://bugs.erlang.org/browse/ERL-193

---

```
This issue became apparent while debugging memory leak in one of applications which is using mochiweb.
I have managed to reduce the issue to a simple gen_tcp socket manipulation so hopefully it will be easy to pinpoint.

{code:erlang}
% erl 1
%% {exit_on_close, false} is crucial for this issue to surface.
f(ListenSocket), {ok, ListenSocket} = gen_tcp:listen(6969, [{exit_on_close, false}]).
f(AcceptSocket), {ok, AcceptSocket} = gen_tcp:accept(ListenSocket).
ok = gen_tcp:send(AcceptSocket, <<0:(10*1024*1024*8)>>).
% erl 2
%% We use {active, false} to prevent receiving the data in mailbox.
f(Socket), {ok, Socket} = gen_tcp:connect({127, 0, 0, 1}, 6969, [{active, false}]).
ok = gen_tcp:close(Socket).
% erl 1
%% We can notice that socket is still there with send_pend and queue_size, wtf?
inet:getstat(AcceptSocket).
{code}

Transcript from Erlang shell #1
{code:erlang}
Erlang/OTP 18 [erts-7.3.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V7.3.1  (abort with ^G)
1> % erl 1
1> %% {exit_on_close, false} is crucial for this issue to surface.
1> f(ListenSocket), {ok, ListenSocket} = gen_tcp:listen(6969, [{exit_on_close, false}]).
{ok,#Port<0.537>}
2> f(AcceptSocket), {ok, AcceptSocket} = gen_tcp:accept(ListenSocket).
{ok,#Port<0.548>}
3> ok = gen_tcp:send(AcceptSocket, <<0:(10*1024*1024*8)>>).
ok
4> % erl 1
4> %% We can notice that socket is still there with send_pend and queue_size, wtf?
4> inet:getstat(AcceptSocket).
{ok,[{recv_oct,0},
     {recv_cnt,0},
     {recv_max,0},
     {recv_avg,0},
     {recv_dvi,0},
     {send_oct,10485760},
     {send_cnt,1},
     {send_max,10485760},
     {send_avg,10485760},
     {send_pend,9577800}]}
{code}

Transcript from Erlang shell #2
{code:erlang}
Erlang/OTP 18 [erts-7.3.1] [source] [64-bit] [smp:4:4] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V7.3.1  (abort with ^G)
1> % erl 2
1> %% We use {active, false} to prevent receiving the data in mailbox.
1> f(Socket), {ok, Socket} = gen_tcp:connect({127, 0, 0, 1}, 6969, [{active, false}]).
{ok,#Port<0.537>}
2> ok = gen_tcp:close(Socket).
ok
{code}

Proof that there is no socket in the system:
{code:bash}
jczarniecki@al-car-06174:~$ lsof -n -iTCP:6969
COMMAND    PID        USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
beam.smp 56033 jczarniecki   24u  IPv4 0x4a433dda029d57cb      0t0  TCP *:acmsoda (LISTEN)
{code}

So basically what's happening here is despite socket is physically gone from system TCP stack, FD is gone and no longer available, the port still lingers while there is no chance that send neither read operation will succeed ever again.

I suppose it's somewhere in erts/emulator/drivers/common/inet_drv.c code...

Kind regards,
Jakub Czarniecki
```
